ST6RI-682 Implement loading of models from repository

TheKorpos opened the pull request:
This PR adds basic model download capability using the standard API.

### Overview
The model download functionality can be used
1. In Jupyter using `magic` commands
2. In Eclipse IDE as a "right click" action on projects
3. As command line java application

In Jupyter, the downloaded models are stored in memory and can be used throughout the work session. The Eclipse and command-line versions serialize the model into separate XMI files (`.sysmlx`). In the Eclipse IDE `.kermlx` and `.sysmlx` files are now added to the Xtext index so their content can be referenced using the textual syntax.

### Common

- `.kermlx` parsing and indexing: 
   - Added `org.omg.kerml.xtext.xmi`, `org.omg.kerml.xtext.ui.xmi`
- `.sysmlx` parsing and indexing: 
   - Added `org.omg.sysml.xtext.xmi`, `org.omg.sysml.xtext.ui.xmi`
- APIs for model download, JSON to EMF transformation
   - `org.omg.sysml.util.repository.ProjectRepository`: repository access, download
   - `org.omg.sysml.xtext.ui.util.RepositoryContentFetcher`: JSON to EMF transformation, XMI serialization (`.sysmlx`)

### Jupyter

New magic commands:

```
%projects

Print the name and identifier of all projects in the repository.
```

```
%load [--id=<PROJECT ID>] [--name=<NAME>] [--branchid=<BRANCH ID>] [--branch=<BRANCH_NAME>] [<NAME>]

Download previously published models from a project in the repository. (Use %projects to view repository contents.)
Named elements of the downloaded models may then be referenced by models in the notebook.
<NAME> is the full name of the project.
If <PROJECT ID> is given, then the project with that UUID is loaded. In this case, the <NAME> must not be given.
If <BRANCH NAME> or <BRANCH ID> is given, then the model is loaded from this branch of the project.
If no <BRANCH NAME> or <BRANCH ID> is given, the default branch is used.
If <BRANCH ID> is given, then a <BRANCH NAME> must not be given.
```

### Eclipse

- Added `SysMLv2` menu to the popup menu for projects, with the following options:
  - `Generate Library Index`: moved under `SysMLv2` menu.
  - `Pull model from API`: download models from a repository and saves in `.sysmlx`
     - Requires that  `.settings/org.omg.sysml.remote.properties` exists with the following contents:
       ```
       base.url=repository url
       remote.projectId=project id
       remote.branchId=branch id
       ```
       (`remote.branchId` is optional; if not provided, the default branch is used)

**NOTE:** 
A specific [tester](https://github.com/Systems-Modeling/SysML-v2-Pilot-Implementation/pull/616/commits/791f814fce088340af150e3d885e444cccd55e7a) was implemented to hide the "Pull model from API" action from projects that don't have the required properties file (`org.omg.sysml.remote.properties`). However, the `sysml.ui plugin` must be activated for this tester to take effect. The plugin activation happens the first time the sysml editor is opened.

### Command Line Application
- `org.omg.sysml.xtext.util.SysMLRepositoryLoadUtil`
  - arguments: 
    `-p, --project`: name of the project in the repository
    `-l, --library`: location of the standard library
    `-t, --target`: target location for saved resources
    `-b, --base` (optional): base URL of the repository, default: `http://localhost:9000`

seidewitz reviewed (changes requested):
- Instead of `%publications` and `%download` use `%projects` and `%load`.
- The `%load` command (and utility) should have an option to use the project ID instead of the name.

hpdekoning commented:
@TheKorpos 
Hi Laszlo, just happened to scan through this PR, and have a minor comment w.r.t to long-names for options.
> org.omg.sysml.xtext.util.SysMLRepositoryLoadUtil has been added
        arguments:
        -p, -project: name of the project in the repository
        -l, -library: location of the standard library
        -t, -target: target location for saved resources
        -b, -base (optional): URL of the repository, default: http://localhost:9000/
> as for the utility project name is now passed with -n <PROJECT NAME> or -name <PROJECT NAME>.
alternatively -id <PROJECT ID> can be used

In my experience, most open source tools use the GNU convention for command line options is to use single hyphen like `-n` for single letter (short-name) options and double hyphen like `--name` for fully spelled out long-name options (two or more characters), which would also be in line with the magic commands. From documentation https://commons.apache.org/proper/commons-cli/ the `org.apache.commons.cli` package does support this. Or perhaps there is another reason not to stay with GNU conventions?

TheKorpos commented:
> @TheKorpos Hi Laszlo, just happened to scan through this PR, and have a minor comment w.r.t to long-names for options.
> 
> > org.omg.sysml.xtext.util.SysMLRepositoryLoadUtil has been added
> > arguments:
> > -p, -project: name of the project in the repository
> > -l, -library: location of the standard library
> > -t, -target: target location for saved resources
> > -b, -base (optional): URL of the repository, default: http://localhost:9000/
> > as for the utility project name is now passed with -n  or -name .
> > alternatively -id  can be used
> 
> In my experience, most open source tools use the GNU convention for command line options is to use single hyphen like `-n` for single letter (short-name) options and double hyphen like `--name` for fully spelled out long-name options (two or more characters), which would also be in line with the magic commands. From documentation https://commons.apache.org/proper/commons-cli/ the `org.apache.commons.cli` package does support this. Or perhaps there is another reason not to stay with GNU conventions?

@hpdekoning 
Thanks for the pointing this out!

`org.apache.commons.cli ` does have a `GNUParser ` but it's marked deprecated. I used the suggested `DefaultParser` instead. This parser allows long-name options with both single and double hyphen. Intrestigely the dedicated GNUParser allows this behavior as well it seems.

There is a `allowPartialMatching ` mode too. This is true by default. This allows partially spelled long-names as long as there is only one matching option. I think I should turn this off as it makes `--n=something` a valid option.

I'm not sure that allowing single hyphen long-names is a huge issue as long as it also accepts the GNU argument syntax.

ujhelyiz reviewed (commented):
In general this looks fine, but it is possible that some comment added to #633 might be better handled here.

seidewitz reviewed (changes requested):
- Update the output from the `%projects` command so it has the following format:
   ```
   API base path: <base path URL>

   Project <project name> (<project UUID>)
   ...
   ```
- The `%load` command should provide clearer feedback on intermediate progress.
- The returned success result from the `%load` command should have the form:
   ```
   Loaded Project <project name> (<project UUID>)
   ```
- I question whether we should include the `--branch` option as "experimental". Either we should include it and tests, with whatever limitations we currently have, or we should just not include it in this release.

TheKorpos commented:
@seidewitz

I changed the output of `%projects` as requested.
 
> * The `%load` command should provide clearer feedback on intermediate progress.
> * The returned success result from the `%load` command should have the form:
>   ```
>   Loaded Project <project name> (<project UUID>)
>   ```

I changed the feedback. Now it only displays the two long running processes: the library UUID collection and the download/model creation. The UUID collection only happens on the first load so it's only displayed once when it happens.

This is how the output looks:
```
Caching library UUIDs...
Downloading model...
Loaded Project <project name> (project UUID)
```
> * I question whether we should include the `--branch` option as "experimental". Either we should include it and tests, with whatever limitations we currently have, or we should just not include it in this release.

The branch upload/download works with the pilot implementation but I'm not sure how much value it adds with the current limitations. Currently uploading to a branch (other than main) is basically the same as uploading to a seperate project.

seidewitz commented:
When I try to pull a model from the API in Eclipse using the following settings:
```
base.url=https://sysml2-dev.intercax.com:9000
remote.projectId=02336c2a-4c7b-4518-b980-addb71d9d378
```
I get the following exception:
```
java.util.ConcurrentModificationException
	at org.eclipse.emf.common.util.AbstractEList$EIterator.checkModCount(AbstractEList.java:751)
	at org.eclipse.emf.common.util.AbstractEList$EIterator.doNext(AbstractEList.java:699)
	at org.eclipse.emf.common.util.AbstractEList$EIterator.next(AbstractEList.java:685)
	at org.eclipse.xtext.resource.SynchronizedXtextResourceSet$1$SynchronizedEIterator.next(SynchronizedXtextResourceSet.java:196)
	at org.eclipse.xtext.resource.SynchronizedXtextResourceSet$1$SynchronizedEIterator.next(SynchronizedXtextResourceSet.java:1)
	at org.omg.sysml.util.repository.EObjectUUIDTracker.trackUUIDSFromResources(EObjectUUIDTracker.java:69)
	at org.omg.sysml.util.repository.EObjectUUIDTracker.trackLibraryUUIDs(EObjectUUIDTracker.java:55)
	at org.omg.sysml.xtext.ui.handlers.PullRepositoryProject.tryPullRemoteFor(PullRepositoryProject.java:130)
	at org.omg.sysml.xtext.ui.handlers.PullRepositoryProject$1.runInWorkspace(PullRepositoryProject.java:97)
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:43)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

TheKorpos commented:
Hi @seidewitz

> java.util.ConcurrentModificationException
> 	at org.eclipse.emf.common.util.AbstractEList$EIterator.checkModCount(AbstractEList.java:751)
> 	at org.eclipse.emf.common.util.AbstractEList$EIterator.doNext(AbstractEList.java:699)
> 	at org.eclipse.emf.common.util.AbstractEList$EIterator.next(AbstractEList.java:685)
> 	at org.eclipse.xtext.resource.SynchronizedXtextResourceSet$1$SynchronizedEIterator.next(SynchronizedXtextResourceSet.java:196)
> 	at org.eclipse.xtext.resource.SynchronizedXtextResourceSet$1$SynchronizedEIterator.next(SynchronizedXtextResourceSet.java:1)
> 	at org.omg.sysml.util.repository.EObjectUUIDTracker.trackUUIDSFromResources(EObjectUUIDTracker.java:69)
> 	at org.omg.sysml.util.repository.EObjectUUIDTracker.trackLibraryUUIDs(EObjectUUIDTracker.java:55)
> 	at org.omg.sysml.xtext.ui.handlers.PullRepositoryProject.tryPullRemoteFor(PullRepositoryProject.java:130)
> 	at org.omg.sysml.xtext.ui.handlers.PullRepositoryProject$1.runInWorkspace(PullRepositoryProject.java:97)
> 	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:43)
> 	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)

I couldn't reproduce this issue. I do get some erros sometimes when the library is not properly built but this time around I cound't break it enough to get an error like this.

Https doens't work right now but I can reach the repo through http.

seidewitz commented:
> I couldn't reproduce this issue. I do get some erros sometimes when the library is not properly built but this time around I couldn't break it enough to get an error like this.
> 
> Https doesn't work right now but I can reach the repo through http.

Oops, I knew that the Intercax repo deployment didn't work with https. The pull from repository works when I change the URL to http.

seidewitz commented:
I updated `EMFModelDelta` (on the ST6RI-178 branch, PR #633) to use the declared name of a root element (if there is one) as the name of the `.sysmlx` file created when pulling a project from the repository.

TheKorpos commented:
I've moved the .gitignores to the new packages (of course I mistyped the issue number...) and updated the package exports in the manifest

seidewitz commented:
> I've moved the .gitignores to the new packages (of course I mistyped the issue number...) and updated the package exports in the manifest

It turns out that I actually had already added the new package exports to the manifests, but I had somehow forgotten to delete the old packages that were replaced. While I don't think having redundant exports causes any real problems, I just removed the redundancies for clarity.
